### PR TITLE
Feature/vlt packages

### PR DESCRIPTION
**Pull request by ist-rw**

### CODE REVIEW

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/sync/PackageSyncHandler.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.composum.sling.core.pckgmgr.sync;
+
+/**
+ * Created by rw on 02.12.15.
+ */
+public class PackageSyncHandler {
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/sync/PackageSyncHandler.java]: empty class?

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/sync/PackageSyncService.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.composum.sling.core.pckgmgr.sync;
+
+/**
+ * Created by rw on 02.12.15.
+ */
+public interface PackageSyncService {
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/sync/PackageSyncService.java]: empty interface

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/sync/PackageSyncServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.composum.sling.core.pckgmgr.sync;
+
+/**
+ * Created by rw on 02.12.15.
+ */
+public class PackageSyncServiceImpl implements PackageSyncService {
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/sync/PackageSyncServiceImpl.java]: empty class?

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageProgressTracker.java [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+        public final String path;
+        public final String message;
+        public final String error;
+
+        public final boolean actionOnly;
+        public final boolean errorDetected;
+
+        public Item(Mode mode, String action, String path, Exception ex) {
+            this.action = action;
+            if (Mode.TEXT.equals(mode)) {
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageProgressTracker.java]: enum comparison can be done wit ==

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageProgressTracker.java [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+            } else {
+                writer.append(PLAIN_TEXT_SHORT_ACTION_SPACE);
+            }
+            writer.append(item.path != null ? item.path : item.message);
+            if (item.errorDetected) {
+                writer.append("\n  ! ");
+                writer.append(item.error);
+            }
+            writer.append('\n');
+            if (finalizedIndicator != null && finalizedIndicator.matcher(item.action).matches()) {
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageProgressTracker.java]: item.action might be null

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java [hunk 6]
```diff
@@ -239,107 +268,422 @@ public static String getDownloadUrl(JcrPackage pckg) {
...
+        }
+
+        @Override
+        public void toJson(JsonWriter writer) throws IOException {
+            JsonUtil.jsonMap(writer, this);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            return getName().equals(((TreeItem) other).getName());
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java]: equals contract violated

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java [hunk 7]
```diff
@@ -239,107 +268,422 @@ public static String getDownloadUrl(JcrPackage pckg) {
...
+        public Calendar getLastModified() {
+            Calendar lastModified = PackageUtil.getLastModified(jcrPackage);
+            if (lastModified != null) {
+                return lastModified;
+            }
+            return PackageUtil.getCreated(jcrPackage);
+        }
+
+        @Override
+        public boolean equals(Object other) {
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java]: equals contract violated

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java [hunk 8]
```diff
@@ -239,107 +268,422 @@ public static String getDownloadUrl(JcrPackage pckg) {
...
+
+    //
+    // JSON mapping helpers
+    //
+
+    public static void toJson(JsonWriter writer, JcrPackage jcrPackage,
+                              Map<String, Object> additionalAttributes)
+            throws RepositoryException, IOException {
+        writer.beginObject();
+        Node node = jcrPackage.getNode();
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java]: node is never used

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java [hunk 9]
```diff
@@ -239,107 +268,422 @@ public static String getDownloadUrl(JcrPackage pckg) {
...
+        writer.name("definition");
+        toJson(writer, jcrPackage.getDefinition());
+        JsonUtil.jsonMapEntries(writer, additionalAttributes);
+        writer.endObject();
+    }
+
+    public static void toJson(JsonWriter writer, JcrPackageDefinition definition)
+            throws RepositoryException, IOException {
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/util/PackageUtil.java]: RepositoryException is never thrown

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/view/PackageManagerBean.java [hunk 10]
```diff
@@ -32,4 +36,21 @@ public String getTabType() {
...
+            PackageUtil.TreeNode treeNode = PackageUtil.getTreeNode(request);
+            for (PackageUtil.TreeItem item : treeNode) {
+                if (item instanceof PackageUtil.PackageItem) {
+                    if (((PackageUtil.PackageItem) item).getDefinition() != null) {
+                        items.add((PackageUtil.PackageItem) item);
+                    }
+                }
+            }
+        } catch (RepositoryException e) {
+            e.printStackTrace();
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/view/PackageManagerBean.java]: use log

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/core/PackageServlet.java [hunk 11]
```diff
@@ -215,7 +230,51 @@ public void doIt(SlingHttpServletRequest request, SlingHttpServletResponse respo
...
+
+            JcrPackageManager manager = PackageUtil.createPackageManager(request);
+            JcrPackage jcrPackage = PackageUtil.getJcrPackage(manager, resource);
+
+            if (jcrPackage != null) {
+
+                Property data;
+                Binary binary;
+                InputStream stream;
+                if (jcrPackage != null &&
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/core/PackageServlet.java]: jcrPackage is never null here

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/core/PackageServlet.java [hunk 12]
```diff
@@ -257,38 +316,164 @@ public void doIt(SlingHttpServletRequest request, SlingHttpServletResponse respo
...
+                                      JcrPackage jcrPackage, ImportOptions options)
+                throws RepositoryException, IOException {
+
+            try {
+                jcrPackage.install(options);
+
+                response.setStatus(HttpServletResponse.SC_OK);
+                Writer writer = response.getWriter();
+                writer.append("<repo>");
+//                writer.append("<crx version=\"\" user=\"\" workspace=\"\">");
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/core/PackageServlet.java]: unused code?

File: sling/core/pckgmgr/src/main/resources/root/libs/composum/sling/pckgmgr/jcrpckg/general/pckg-update.jsp [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+                    <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span
+                            aria-hidden="true">&times;</span></button>
+                    <h4 class="modal-title">Change Package Properties</h4>
+                </div>
+                <div class="modal-body">
+                    <div class="messages">
+                        <div class="alert"></div>
+                    </div>
+                    <input name="_charset_" type="hidden" value="UTF-8"/>
+                    <input name="path" type="hidden" value="${pckg.path}">
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/resources/root/libs/composum/sling/pckgmgr/jcrpckg/general/pckg-update.jsp]: where is `pckg` defined?

File: sling/core/pckgmgr/src/main/resources/root/libs/composum/sling/pckgmgr/query.jsp [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+                    <input type="text" class="form-control" placeholder="Search">
+                    <span class="exec glyphicon-triangle-right glyphicon input-group-addon btn btn-default"
+                          title="Execute query..."></span>
+                </div>
+            </cpn:form>
+        </div>
+    </div>
+    <div class="query-result">
+        <table class="table table-striped table-hover table-condensed">
+            <thead>
```
Comment (mzeibig) [inline sling/core/pckgmgr/src/main/resources/root/libs/composum/sling/pckgmgr/query.jsp]: missing `<tr>`

File: sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/core/PackageServlet.java (hunk 12 above)
Comment (ist-rw) [inline sling/core/pckgmgr/src/main/java/com/composum/sling/core/pckgmgr/core/PackageServlet.java]: the typical answer from CRX as comment with the necessary code uncommented

**Pull request merged**